src/connection/activate.ts

```diff
@@ -0,0 +1,94 @@
+import { sendEvent } from 'components/analytics'
+import { Connection } from 'connection'
+import { atom, useAtom } from 'jotai'
+import { useCallback } from 'react'
+import { useAppDispatch } from 'state/hooks'
+import { updateSelectedWallet } from 'state/user/reducer'
+
+import { didUserReject } from './utils'
+
+type PendingConnectionState = {
+  connection: Connection | undefined
+  error: any
+}
+
+const pendingConnectionAtom = atom<PendingConnectionState>({ connection: undefined, error: undefined })
+
+export function useActivateConnection() {
+  const dispatch = useAppDispatch()
+
+  const [pending, setPending] = useAtom(pendingConnectionAtom)
+
+  const updatePendingConnection = useCallback(
+    (connection: Connection | undefined) => {
+      // log selected wallet
+      if (connection !== undefined) {
+        sendEvent({
+          category: 'Wallet',
+          action: 'Change Wallet',
+          label: connection.type,
+        })
+      }
+      // Setting a new connection clears any pending errors
+      setPending({ connection, error: undefined })
+    },
+    [setPending]
+  )
+
+  const updatePendingError = useCallback(
+    (error: any) => {
+      setPending(({ connection }) => {
+        if (error !== undefined && connection !== undefined) {
+          sendAnalyticsEvent(InterfaceEventName.WALLET_CONNECT_TXN_COMPLETED, {
+            result: WalletConnectionResult.FAILED,
+            wallet_type: connection.getName(),
+          })
+        }
+
+        return { connection, error }
+      })
+    },
+    [setPending]
+  )
+
+  const cancelActivation = useCallback(async () => {
+    await pending.connection?.connector.deactivate?.()
+    updatePendingConnection(undefined)
+  }, [pending, updatePendingConnection])
+
+  const tryActivation = useCallback(
+    async (connection: Connection, onSuccess: () => void) => {
+      // Skips wallet connection if the connection should override the default behavior, i.e. install metamask or launch coinbase app
+      if (connection.overrideActivate?.()) return
+
+      try {
+        updatePendingConnection(connection)
+        await connection.connector.activate()
+
+        console.debug(`connection activated: ${connection.getName()}`)
+        dispatch(updateSelectedWallet({ wallet: connection.type }))
+
+        // Clears pending connection state
+        updatePendingConnection(undefined)
+
+        onSuccess()
+      } catch (error) {
+        // TODO(WEB-3162): re-add special treatment for already-pending injected errors
+        console.debug(`web3-react connection error: ${JSON.stringify(error)}`)
+
+        // Gracefully handles errors from the user rejecting a connection attempt
+        if (didUserReject(connection, error)) {
+          updatePendingConnection(undefined)
+          return
+        }
+
+        updatePendingError(error)
+      }
+    },
+    [dispatch, updatePendingConnection, updatePendingError]
+  )
+
+  return { pending, tryActivation, cancelActivation }
```

Comment:
I think it would be a bit better if this exported React Query-like API, that is: `isPending`, `isError`, `error`, `tryActivation` and `cancelActivation` instead of having `pending.error` under it.

I would be open to naming it: `isPendingActivation`, `isActivationError` and `activationError` to avoid renaming later on the call-site, so it's more explicit.

Reply:
I think adding those might feel too boilerplate-y for this hook's use-case & I'm not sure the two components that use this hook will benefit from shaping this more like an api